### PR TITLE
zmq: mempool notifications

### PR DESCRIPTION
**Pull request by laanwj**
Add notifications when transactions enter or leave the mempool.

These can be enabled with `zmqpubmempool`:
- `mempooladded`: a transaction was added to the mempool
- `mempoolremoved`: a transaction was removed from the mempool

This allows third-party software to keep track of what is in the mempool in real time. For example a mempool monitor I'm working on:
schermafdruk van 2016-03-27 17-42-16
It also makes it possible to keep statistics on why transactions disappear from the mempool.

Full documentation: https://github.com/bitcoin/bitcoin/blob/325afe68381d59f1b95a690927c4a8ea886ac791/doc/zmq.md#mempooladd

See individual commits for details.

Current issues:
- Warning `NotifyEntryAdded: WARNING: zmq send failed with code -1` spam in log if "-zmqpubmempool" is not enabled.

### CODE REVIEW

**Comment by promag** [comment]
I was planning to submit a pull request to change this kind of *notifications*. The current solution is not very elegant and doesn't scale.

So if you don't mind I'll try to explain here the idea using this example:

Usage: `-notify "mempool ! zmqpub address=tcp://127.0.0.1:28332"`

Here the notify argument creates one notification *pipeline* composed by one *source* of events (mempool) and one *sink* of events (zmqpub), connected with the operator `!`.

**Comment by dcousens** [comment]
concept ACK

**Comment by laanwj** [comment]
@promag I'm all for making it more flexible, but what you're proposing sounds like the gstreamer syntax and that's IMO even harder to use than the current arguments. I've needed to do a few media pipelines in my PhD research but never managed to really master it.

Our basic idea with the ZMQ notifications is that **a** notification system exists. From there on everything can be handled by external software. I don't want too much notification-related complexity in core.

Aside, this is the wrong place to discuss this. Let's focus on the software change here. Please create a new issue.

**Comment by promag** [comment]
@laanwj yes gstreamer all over the comment. I'll move the discussion to a new issue.

utACK.

File: src/init.cpp [hunk 1]
```diff
@@ -1230,7 +1230,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
         AddOneShot(strDest);
 
 #if ENABLE_ZMQ
-    pzmqNotificationInterface = CZMQNotificationInterface::CreateWithArguments(mapArgs);
+    pzmqNotificationInterface = CZMQNotificationInterface::CreateWithArguments(mapArgs, &mempool);
```
Comment (promag) [inline src/init.cpp]: Feels a bit hacky.

Comment (laanwj) [inline src/init.cpp]: Better suggestions?
As I see ti there are two alternatives:
- use the global mempool object (ugly)
- pass the mempool in (slightly les ugly)

**Comment by jonasschnelli** [comment]
Awesome!
Nice work! I hope you have also plans to opensource the ncurses UI.

Some ideas / conceptual questions:
- How could a listener keep track of the overall mempool size without listening to a fresh start (listening from the beginning when the mempool was at a size of 0), also there is a risk of a delta in case you missed a notification in case you calculate the mempool overall size on the "listeners side".

- would a notification after a significant (every X MB, every X transaction added/removed) in the mempool size or structure make sense?

- Adding the total mempool size in every mempool notification is probably to much?

File: src/init.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/init.cpp]: I think this implementation approach is reasonable and I also don't see a more elegant way to do this.

**Comment by jonasschnelli** [comment]
Code Review utACK.

File: src/init.cpp (hunk 1 above)
Comment (promag) [inline src/init.cpp]: Then I would rename `CreateWithArguments` to just `Create` and move the following from src/init.cpp to `Create`.

```
if (pzmqNotificationInterface) {
    RegisterValidationInterface(pzmqNotificationInterface);
}
```

**Comment by promag** [comment]
> Adding the total mempool size in every mempool notification is probably to much?

:+1:

**Comment by laanwj** [comment]
> How could a listener keep track of the overall mempool size without listening to a fresh start (listening from the beginning when the mempool was at a size of 0), also there is a risk of a delta in case you missed a notification in case you calculate the mempool overall size on the "listeners side".

What if you use the following sequence:
- Start listening (queue up events)
- Request `getrawmempool`
- Start applying events

You may have some duplicates in the first run, e.g., adds which already show up in the mempool requested, removes which were already removed, but I think due to the nature of the events (just add and remove) the overall state will be the same?

>  there is a risk of a delta in case you missed a notification

Missing notifications are a possibiltiy with zmq. But I'd say this is a matter of sizing buffers appropriately, and dedicating a thread to listening to zmq requests.

To reliably detect missing notifications we could add sequence numbers to the events, as I think zmq guarantees that events from a single source are delivered in order.

Then again if we do this we probably want it for all the event types...

> Adding the total mempool size in every mempool notification is probably to much?

I thought about sending a generic stats packet with every mempool event, but as you can receive tons of them I don't want to make the individual events too big and include things you can compute yourself. But no problems with just the size / # transactions.

**Comment by jonasschnelli** [comment]
Tested ACK 4e3039b41946ad7bbb1aff6b6df5b1e83348c942
I also think the `addNotifier()`-refactor makes it much cleaner.

**Comment by paveljanik** [comment]
lightly tested ACK https://github.com/bitcoin/bitcoin/commit/4e3039b41946ad7bbb1aff6b6df5b1e83348c942

File: src/init.cpp (hunk 1 above)
Comment (laanwj) [inline src/init.cpp]: I agree, will move that line.

File: src/txmempool.h [hunk 2]
```diff
@@ -319,6 +321,18 @@ class CInPoint
     size_t DynamicMemoryUsage() const { return 0; }
 };
 
+/** Reason why a transaction was removed from the mempool,
+ * this is passed to the notification signal.
+ */
+enum MemPoolRemovalReason {
```
Comment (sipa) [inline src/txmempool.h]: This enum seems to be part of the ZMQ API, so we'll probably want some stability across versions. Things like MPR_EXPIRY and MPR_SIZELIMIT seem quite specific to the existing implementation.

File: src/zmq/zmqpublishnotifier.cpp [hunk 3]
```diff
@@ -171,3 +177,41 @@ bool CZMQPublishRawTransactionNotifier::NotifyTransaction(const CTransaction &tr
...
+    if (rc != 0)
+        LogPrintf("%s: WARNING: zmq send failed with code %i\n", __func__, rc);
+}
+
+void CZMQPublishMempoolNotifier::NotifyEntryRemoved(const CTxMemPoolEntry &entry, MemPoolRemovalReason reason)
+{
+    LogPrint("zmq", "zmq: mempool entry removed: %s, reason %i\n", entry.GetTx().GetHash().ToString(), (int)reason);
+    uint8_t data[32 + 1];
+    WriteHash(&data[0], entry.GetTx().GetHash());
+    data[32] = reason;
```
Comment (sipa) [inline src/zmq/zmqpublishnotifier.cpp]: I would prefer it if the value of the (internal) MemPoolRemovalReason did not leak into the ZMQ protocol, though that can be fixed later by introducing a new enum in the ZMQ definition.

Comment (laanwj) [inline src/zmq/zmqpublishnotifier.cpp]: Yes, I'll add another enum here and add a mapping.

**Comment by laanwj** [comment]
Ok added two commits:
- The first adds a (currently, one to one) mapping between the mempool reasons from the mempool and those on the ZMQ interface.
- The second is a slight refactoring to make all ZMQ publish notifiers register (and unregister) themselves for the events that they want to receive, instead of relying on the events being propagated manually though `CZMQNotificationInterface` @promag

**Comment by laanwj** [comment]
Rebased after #7762

**Comment by MarcoFalke** [comment]
Travis failure seems unrelated

```
Your test run exceeded 50 minutes. 
One possible solution is to split up your test run.

**Comment by laanwj** [comment]
Rebased, and changed the enums (both for the mempool and for the zmq protocol) to c++11 enums.

**Comment by paveljanik** [comment]
lightly tested ACK https://github.com/bitcoin/bitcoin/pull/7753/commits/1bbe3660ad9a375e2d2bb8c7fa8f38132774b656

**Comment by sipa** [comment]
Needs rebase.

**Comment by jonasschnelli** [comment]
Still needs rebase.

**Comment by jmcorgan** [comment]
ACK.  I have a rebased version of this branch at 'jmcorgan/2016_03_zmq_mempool_notifications'.

**Comment by jonasschnelli** [comment]
@jmcorgan feel free to open a PR of your rebases version and take over the lead.

**Comment by jmcorgan** [comment]
Continued in #8549

**Comment by laanwj** [comment]
Closing in favor of #8549

**Pull request closed without merging**